plugin-server/src/main/ingestion-queues/kafka-queue.ts

```diff
@@ -137,15 +140,24 @@ export class KafkaQueue {
         return await startPromise
     }
 
-    async bufferSleep(sleepMs: number, partition: number): Promise<void> {
+    async bufferSleep(
+        sleepMs: number,
+        partition: number,
+        offset: string,
+        heartbeat: () => Promise<void>
+    ): Promise<void> {
+        await this.pause(this.bufferTopic, partition)
+        const sleepHeartbeatInterval = setInterval(async () => {
+            await heartbeat() // Send a heartbeat once a second so that the broker knows we're still alive
+        }, 1000)
         this.sleepTimeout = setTimeout(() => {
             if (this.sleepTimeout) {
                 clearTimeout(this.sleepTimeout)
             }
+            clearInterval(sleepHeartbeatInterval)
+            this.consumer.seek({ topic: this.bufferTopic, partition, offset })
```

Comment:
This approach has a new problem: it is not robust against crashes.

We're relying on memory in a single server to tell us where we should consume from next. If a deploy happens while we're sleeping, we lose this context and as a result lose events.

We seem to be getting hackier and hackier here. (the buffer itself is a _very hacky_ concept, I know). What was the problem with throwing again?

Otherwise, I actually found your previous solution to be clean and easy to follow, and it can work if have two consumers

Reply:
We won't lose events, because we `resolveOffset()` of every message that's ingested, and then `commitOffsetsIfNecessary()` is called _before_ this `.bufferSleep()` method – so if the plugin server crashes, processing will resume from the last processed message. Note that offsets of messages after the cuttoff message are not resolved.
If autocommit decides not commit at `commitOffsetsIfNecessary()` and _then_ the plugin server crashes, some events might be processed _more_ than once, but that's also exactly how the base queue works and doesn't seem to be an issue.
I think I haven't clearly stated what's the issue with the error approach. It simply doesn't work. Throwing errors results in the consumer shutting down. It also results in a lot of log spam. Shoehorning error handling doesn't seem to be the right way here, but this solution suggested by the KafkaJS contributor seems pretty solid.

Reply:
Thanks for the explanation!

I'm a bit confused though - if it's all committed correctly, why is seeking needed at all?

Reply:
Added a comment explaining this:
We need to seek so that after resuming we start exactly where we left off (with the cutoff message) instead of skipping over messages left for later in the last batch.
Normally KafkaJS would go [messages 1-5] → [messages 6-9] (batch sizes not constant) regardless of what offsets are resolved, we need to tell it to chunk things differently if we want to process part of the batch later. This way we get: [messages 1-5] oops, we need to wait until 4 and 5 can be processed! → pause and wait → seek back to 4 → resume → [messages 4-11]

Reply:
Are you sure it ignores committed offsets when picking up the next batch? That seems odd.

Either way, if you're telling me we're committing 4 and only using `seek` to alter the in-memory behavior of the consumer then this might be fine

Reply:
That's exactly the case, yeah.

Reply:
Cool, let me do a final pass then